test/Nerdbank.MessagePack.Tests/MessagePackStreamingReaderTests.cs

```diff
@@ -0,0 +1,113 @@
+		Assert.Equal(3, count);
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out bool boolean));
+		Assert.False(boolean);
+		Assert.Equal(DecodeResult.InsufficientBuffer, incompleteReader.TryRead(out boolean));
+		incompleteReader = new(await incompleteReader.ReplenishBufferAsync());
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out boolean));
+		Assert.True(boolean);
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out boolean));
+		Assert.False(boolean);
+
+		Assert.Equal(DecodeResult.EmptyBuffer, incompleteReader.TryReadNil());
+	}
+
+	[Fact]
+	public async Task ReplenishBufferAsync_AddsMoreBytes_ThenCompletes()
+	{
+		// Arrange the reader to have an incomplete buffer and that upon request it will get the rest of it.
+		int callCount = 0;
+		MessagePackStreamingReader incompleteReader = new(
+			ArrayOf3Bools.Slice(0, 2),
+			(_, pos, ct) => new(new ReadResult(ArrayOf3Bools.Slice(pos), false, isCompleted: ++callCount > 1)),
+			null);
+
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryReadArrayHeader(out int count));
+		Assert.Equal(3, count);
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out bool boolean));
+		Assert.False(boolean);
+		Assert.Equal(DecodeResult.InsufficientBuffer, incompleteReader.TryRead(out boolean));
+		incompleteReader = new(await incompleteReader.ReplenishBufferAsync());
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out boolean));
+		Assert.True(boolean);
+		Assert.Equal(DecodeResult.Success, incompleteReader.TryRead(out boolean));
+		Assert.False(boolean);
+
+		Assert.Equal(DecodeResult.InsufficientBuffer, incompleteReader.TryReadNil());
+		incompleteReader = new(await incompleteReader.ReplenishBufferAsync());
+		Assert.Equal(DecodeResult.EmptyBuffer, incompleteReader.TryReadNil());
+	}
+
+	private static ReadOnlySequence<byte> CreateMsgPackArrayOf3Bools()
+	{
+		Sequence<byte> seq = new();
+		MessagePackWriter writer = new(seq);
+		writer.WriteArrayHeader(3);
+		writer.Write(false);
+		writer.Write(true);
+		writer.Write(false);
+		writer.Flush();
+
+		return seq;
+	}
+
+	// TODO: Remove/move this sample
+	[MessagePackConverter(typeof(SomeTypeConverter))]
+	record SomeType(string Message, int Number);
+
+	class SomeTypeConverter
+	{
+		internal async ValueTask<(SomeType, MessagePackStreamingReader.BufferRefresh)> Read(MessagePackStreamingReader.BufferRefresh readerInputs)
+		{
+			MessagePackStreamingReader reader = new(readerInputs);
+
+			while (reader.TryReadArrayHeader(out int count).NeedsMoreBytes())
+			{
+				reader = new(await reader.ReplenishBufferAsync());
+			}
+
+			string? str;
+			while (reader.TryRead(out str).NeedsMoreBytes())
+			{
+				reader = new(await reader.ReplenishBufferAsync());
+			}
+
+			int num;
+			while (reader.TryRead(out num).NeedsMoreBytes())
+			{
+				reader = new(await reader.ReplenishBufferAsync());
+			}
+
+			return (new SomeType(str, num), reader.GetExchangeInfo());
+		}
```

Comment:
@eiriktsarpalis This is as good as I have been able to imagine so far. Do you think we can do any better?

Reply:
What are the drawbacks of making `BufferRefresh` a class? If allocation is the main reason instances can always be pooled by the serializer (we do this fairly effectively with `Utf8JsonWriter` in STJ).

Reply:
Yes, it's primarily to avoid an allocation or the cost of pooling.
What would the benefits be of making it a class?

Reply:
That would certainly be a better experience. And that's actually the *current* (pre-PR) design. The problem with that is that classes cannot store `ref struct` fields, and there is no ReadOnlyMemory-based `SequenceReader` class (which seems like a serious missing feature of .NET).
Accessing ReadOnlySequence frequently is *slow*. 

If we pass around a class instead of a ref struct, and then each async Read converter method uses the class to obtain a ref struct based on the existing buffer, I guess that could work.

Reply:
I suppose my question then becomes can we get rid of ref structs altogether assuming there is an efficient pooling strategy for reader objects?

Reply:
The obvious downside of such a move being that you can't use wrap a reader around a user-provided ROS<byte>

Reply:
Eliminating `ref struct` altogether means that we can't use `SequenceReader<byte>`, and we'll be slicing memory instead of slicing span. I *believe* `ReadOnlySpan<T>` is somehow faster than `ReadOnlyMemory<T>`, though I don't know how significant this is, but I believe that's why `SequenceReader<byte>` is span-based instead of exposing the `ReadOnlyMemory` that it has available due to being based on `ReadOnlySequence<byte>`.